boa_engine/src/module/mod.rs

```diff
@@ -0,0 +1,734 @@
+    /// time.
+    /// - The operation must treat payload as an opaque value to be passed through to
+    /// `FinishLoadingImportedModule`. (can be ignored)
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-HostLoadImportedModule
+    /// [finish]: https://tc39.es/ecma262/#sec-FinishLoadingImportedModule
+    #[allow(clippy::type_complexity)]
+    fn load_imported_module(
+        &self,
+        referrer: Referrer,
+        specifier: JsString,
+        finish_load: Box<dyn FnOnce(JsResult<Module>, &mut Context<'_>)>,
+        context: &mut Context<'_>,
+    );
+
+    /// Host hooks [`HostGetImportMetaProperties ( moduleRecord )`][meta] and
+    /// [`HostFinalizeImportMeta ( importMeta, moduleRecord )`][final].
+    ///
+    /// This unifies both APIs into a single hook that can be overriden on both cases.
+    /// The most common usage is to add properties to `import_meta` and return, but this also
+    /// allows modifying the import meta object in more exotic ways before exposing it to ECMAScript
+    /// code.
+    ///
+    /// The default implementation of `HostGetImportMetaProperties` is to return a new empty List.
+    ///
+    /// [meta]: https://tc39.es/ecma262/#sec-hostgetimportmetaproperties
+    /// [final]: https://tc39.es/ecma262/#sec-hostfinalizeimportmeta
+    fn init_import_meta(
+        &self,
+        _import_meta: JsObject,
+        _module: Module,
+        _context: &mut Context<'_>,
+    ) {
+    }
+}
+
+/// A simple module loader that loads modules relative to a root path.
+#[derive(Debug)]
+pub struct SimpleModuleLoader {
+    root: PathBuf,
+    module_map: GcRefCell<FxHashMap<PathBuf, Module>>,
+}
+
+impl SimpleModuleLoader {
+    /// Creates a new `SimpleModuleLoader` from a root module path.
+    pub fn new<P: AsRef<Path>>(root: P) -> JsResult<Self> {
+        let root = root.as_ref();
+        let absolute = root.canonicalize().map_err(|e| {
+            JsNativeError::typ()
+                .with_message(format!("could not set module root `{}`", root.display()))
+                .with_cause(JsError::from_opaque(js_string!(e.to_string()).into()))
+        })?;
+        Ok(Self {
+            root: absolute,
+            module_map: GcRefCell::default(),
+        })
+    }
+
+    /// Inserts a new module onto the module map.
+    #[inline]
+    pub fn insert(&self, path: PathBuf, module: Module) {
+        self.module_map.borrow_mut().insert(path, module);
+    }
+
+    /// Gets a module from its original path.
+    #[inline]
+    pub fn get(&self, path: &Path) -> Option<Module> {
+        self.module_map.borrow().get(path).cloned()
+    }
+}
+
+impl ModuleLoader for SimpleModuleLoader {
+    fn load_imported_module(
+        &self,
+        _referrer: Referrer,
+        specifier: JsString,
+        finish_load: Box<dyn FnOnce(JsResult<Module>, &mut Context<'_>)>,
+        context: &mut Context<'_>,
+    ) {
+        let result = (|| {
+            let path = specifier
```

Comment:
Because this allows deferring the load to a later point in time, making it possible to concurrently load several modules at the same time. This is deliberate, as shown by the [spec](https://tc39.es/ecma262/#sec-HostLoadImportedModule):

> The [host environment](https://tc39.es/ecma262/#host-environment) must perform [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result), where result is either a [normal completion containing](https://tc39.es/ecma262/#sec-completion-record-specification-type) the loaded [Module Record](https://tc39.es/ecma262/#sec-abstract-module-records) or a [throw completion](https://tc39.es/ecma262/#sec-completion-record-specification-type), either ***synchronously or asynchronously***.

Of course, the ideal implementation would be to have `async fn` in dyn traits available. Unfortunately, static `async fn` in traits are a [`WIP`](https://github.com/rust-lang/rust/issues/91611), and `async fn` in dyn traits aren't [even in the nightly compiler yet](https://github.com/rust-lang/rust/issues/107011), but we should be able to improve the API when those features arrive :)